boards: add usb to stm32f4_disco

metapsychologe opened the pull request:
This patch adds USB capabilities to the stm32f4 discovery board.
As the board has a MicroUSB port that I would like to use as stdout I just added USB.
The needed entry within pinmux.c is already there but USB was missing in the dts file and documentation.

I tested it with the samples/subsys/usb/console.

Signed-off-by: Alexander Polleti <metapsycholo@gmail.com>

erwango commented on boards/arm/stm32f4_disco/Kconfig.defconfig:
```diff
@@ -17,6 +17,16 @@ config UART_STM32_PORT_2
 endif # UART_CONSOLE
 
+if USB_UART_CONSOLE
+
```
Don't you need to enable "USB" as well?

metapsychologe replied:
You are right.
I need USB, USB_DEVICE_STACK but can remove UART_LINE_CTRL.
I'll update the file.

erwango reviewed (changes requested):
Thanks for this proposal.

I guess this change is made to work in other contexts than sample/subsys/usb/console, so I think it would be wise testing in other conditions as this sample defines a set of CONFIG that can hide issue when this sample is not used.

ydamigos reviewed (changes requested):
I know that the virtualcom port is already done in the x86 series but I am wondering if DT is the correct place for it. DT should describe the hardware and virtualcom port is actually software. Having said that, if we will add the virtualcom node in DT at the end, I would prefer to split the PR in two commits. One which enables USB for the board and one for the virtualport.

metapsychologe committed:
boards: add usb to stm32f4_disco

This patch adds USB capabilities to the stm32f4 discovery board.
As the board has a MicroUSB port that I would like to use as stdout I
just added USB. The needed entry within pinmux.c is already there but
USB was missing in the dts file and documentation.

I tested it with the upcomming commit.

Signed-off-by: Alexander Polleti <metapsycholo@gmail.com>

dbkinder reviewed (approved):
doc change LGTM.

erwango commented:
> I know that the virtualcom port is already done in the x86 series but I am wondering if DT is the correct place for it. DT should describe the hardware and virtualcom port is actually software

@ydamigos , this remark makes sense but use of chosen was accepted as possible way to configure a SW element on a HW resource. This is documented here:
http://docs.zephyrproject.org/devices/dts/device_tree.html?highlight=device%20tree#device-tree-vs-kconfig

>  I would prefer to split the PR in two commits. One which enables USB for the board and one for the virtualport.

+1

ydamigos commented:
```
Device Tree’s `chosen` keyword, which allows the user to select a particular instance 
of a hardware device to be used for a concrete purpose by the software. An example
of this is selecting a particular UART for use as the system’s console.
```
@erwango If I understand correct the above, it should be something like `zephyr,console = &usbotg_fs`. `virtualcom` isn't an instance of a hardware device.

erwango commented:
@metapsychologe , would you manage to get things working using  ` zephyr,console = &usbotg_fs` instead of `virtualcom`?

metapsychologe commented:
It's not as simple as that, as the label of usbotg_fs is "OTGFS", while the default label for the virtual com port is "CDC_ACM".
What seems to work is to configure CDC_ACM_PORT_NAME to "OTGFS". But this would lead to 2 devices with the same label, won't it?
Another possibility would be to set CONFIG_UART_CONSOLE_ON_DEV_NAME (that is the device name for console) if CONFIG_USB_UART_CONSOLE was defined.

I would suggest to make this PR only about the USB for the stm32f4_disco and delete the 2nd commit that adds console output to the virtual com port.

According to console output via virtual com port you could probably open an issue and we can discuss on how to handle the issue there. Perhaps we shall also remove the entanglement of the UART console driver and the USB console.

Imho it is a neat feature to use USB as console output as more and more boards have USB with virtual com port as a handy way to communicate with a PC, while UART is manly used with a UART to USB converter inbetween. So it would be nice to make it as beginners friendly as possible to use USB as console, just like `zephyr,console = &usbotg_fs` but this probably is not the place to discuss this.

erwango commented:
> Imho it is a neat feature to use USB as console output as more and more boards have USB with virtual com port as a handy way to communicate with a PC, while UART is manly used with a UART to USB converter inbetween

Sure, btw for boards with embedded ST link (ST nucleo, disco and eval boards) ST Link is providing the VCP with console assigned to SoC UART. Did you try it?
Then you don't need to add anything to board dts.

ydamigos commented:
Boards user manual (https://www.st.com/resource/en/user_manual/dm00039084.pdf) mentions:
```
The ST-LINK/V2-A supports a virtual COM port (VCP) on U2 pin 12 (ST-LINK_TX) and U2 
pin 13 (ST-LINK_RX) but these pins are not connected to the USART of the STM32F407 
microcontroller for mbed support.
Two solutions are possible to connect an STM32F407 USART to the VCP on the PC:
• Using an USART to USB dongle from the market connected for instance to 
STM32F407 USART2 available on connector P1 pin 14 (PA2: USART2_TX) and P1 
pin 13 (PA3: USART2_RX).
• Using flying wires to connect ST-LINK/V2-A virtual COM port (ST-LINK VCP on U2 pin 
12 and 13) to STM32F407 USART2 (PA2 and PA3: P1 pin 14 and 13) as shown in the 
Figure 6  below.
```

metapsychologe commented:
As I'm using the STM32F4DIS-BB expansion for the STM32F4-Discovery that uses one of the recommended pins and my soldering skills a little rusty, I'm not really indending to try this. As the official documentation suggests to use this technique, I suspect it should be working.
So my question remains wether or not I shall keep the virtual COM port within this PR or just leave the USB stuff inside.

erwango commented:
@metapsychologe , you can keep the usb part but remove the virtual com port.

Btw,I think there are jumpers to dissociate ST-Link from default pins on you board, then there should be ST-Link rx/tx slots available on your board that you can link to the USART pin of your choice

metapsychologe commented:
I removed the virtual com port
